OSDOCS-20040: Adds PlanPolicy to RHCL (#112905)
openshift/openshift-docs: modules/proc-rhcl-planpolicy-deploy.adoc

// Module included in the following assemblies:
//
// *deployment/rhcl-planpolicy.adoc

:_mod-docs-content-type: PROCEDURE
[id="proc-rhcl-planpolicy-deploy_{context}"]
=  Deploying the PlanPolicy

[role="_abstract"]

> The `PlanPolicy` resource enforces plan-based rate limits on the HTTPRoute.

Reviewer: ```suggestion
The `PlanPolicy` object enforces plan-based rate limits on the HTTPRoute.
```

> The `PlanPolicy` resource enforces plan-based rate limits on the HTTPRoute.

Reviewer 2: Same question: is the user targeting an HTTPRoute or a Gateway? Both are supported. I would say we should be consistent. For the sake of simplicity, if the authpolicy targets the route, so does the planpolicy. And if the authpolicy targets the gateway,so does the planpolicy.

.Procedure

> . Deploy the `PlanPolicy` resource to enforce plan-based rate limits on the HTTPRoute by by using the following example:

Reviewer: ```suggestion

. Deploy the `PlanPolicy` object to enforce plan-based rate limits on the HTTPRoute by by using the following example:
```
use either object or custom resource throughout

> . Deploy the `PlanPolicy` resource to enforce plan-based rate limits on the HTTPRoute by by using the following example:

Reviewer 2: `s/by by/by/`

+
.Example enforcement of plan-based rate limits

> 

Reviewer: ```suggestion
```
no space

[source,yaml,subs="+quotes"]
----
$ oc apply -f
apiVersion: extensions.kuadrant.io/v1alpha1
kind: PlanPolicy
metadata:
  name:  _<sample_app_planpolicy>_
  namespace: ${KUADRANT_DEVELOPER_NS}
spec:
  targetRef:
    group: gateway.networking.k8s.io
    kind: HTTPRoute
    name: _<sample_app.yaml>_
  plans:
    - tier: gold
      predicate: |
        has(auth.identity) && auth.identity.metadata.annotations["secret.kuadrant.io/plan-id"] == "gold"
      limits:
        daily: 5
    - tier: silver
      predicate: |
        has(auth.identity) && auth.identity.metadata.annotations["secret.kuadrant.io/plan-id"] == "silver"
      limits:
        daily: 2
    - tier: bronze
      predicate: |
        has(auth.identity) && auth.identity.metadata.annotations["secret.kuadrant.io/plan-id"] == "bronze"
      limits:
        daily: 1
----

> Replace `_<sample_app.yaml>_` with the name of your application.

Reviewer: ```suggestion
+
* Replace `_<sample_app.yaml>_` with the name of your application.
```
must have `+`, and this is a bulleted list that should have no line spaces; follow the callout replacement guidance

+
Replace `_<sample_app_planpolicy.yaml>_ with the PlanPolicy for your application.
+

> Replace 'KUADRANT_DEVELOPER_NS' with the namespace for your developer in {product-title}.

Reviewer: ```suggestion
Replace 'KUADRANT_DEVELOPER_NS' with the namespace for your developer in {prodname}.
```
don't use product-title, it does not resolve

+

> Each `predicate` is a CEL expression. The predicates are evaluated in order, and the first matching plan is applied to the authenticated user.

Reviewer 2: Here we are suggesting that the reader use some CEL expressions. I wonder if it
makes sense to add a note making clear that the CEL expressions can be anything
that suits your use case. Since the suggested predicate has
`auth.identity.metadata.annotations["secret.kuadrant.io/plan-id"] == "gold"`,
we need the secrets to be annotated with `secret.kuadrant.io/plan-id`. But it
could be something else. For instance a label, or any other annotation key.  I think it is correct to suggest something that can be used to get something working. My point is that the PlanPolicy CEL expression does not necessarily have to be the one we suggest.
